docs/ds20.md

```diff
@@ -0,0 +1,111 @@
+
+We could squeeze `FWUPDPLU` or `FWUPDFLA` as the `CompatibleID`, but we couldn’t squeeze the plugin name (e.g. `logitech-bulkcontroller`) or the GUID (16 bytes) in an 8 byte `SubCompatibleID`.
+Some non-compliant devices also don’t respond well (as in, they hang, and stop working...) when probing this specific string index.
+
+#### Microsoft OS Descriptors 2.0
+
+The [Microsoft OS Descriptors 2.0 Specification](https://docs.microsoft.com/en-us/windows-hardware/drivers/usbcon/microsoft-os-2-0-descriptors-specification) is more helpful as it defines a new device capability that can return a variable length chunk of data, using a UUID as a key.
+We do need to use the `BOS` *“Binary Object Store”* descriptor, which is only supported for devices using USB 2.1 and newer, although that’s probably the vast majority of devices now.
+Usually the BOS descriptor is used for Wireless USB details, USB 2.0 extensions, SuperSpeed USB connection details and a Container ID.
+
+Usefully, the BOS descriptor does give us is the ability to use the platform capability descriptor, which is `bDevCapabilityType=0x05`. For UUID `D8DD60DF-4589-4CC7-9CD2-659D9E648A9F`, this is identified as a structured blob of data Windows usually uses to put workarounds like the suspend mode of the device and that kind of thing.
+
+We could certainly create a new `bDevCapabilityType` which would allow us to store a binary blob (e.g. `Plugin=foobarbaz\nFlags=QuirkValueHere\n`) but that would be out-of-specification and difficult for vendors to implement.
+
+## Proposal
+
+Devices create a new platform device capability BOS descriptor with `bDevCapabilityType=0x05` and a fwupd-specific UUID, specifically `010aec63-f574-52cd-9dda-2852550d94f0`.
+Using this custom UUID will ensure that Microsoft Windows does not try to parse the capability descriptor as a *Microsoft OS 2.0 descriptor set*.
+
+## Implementation
+
+Create a BOS DS20 descriptor such as:
+
+    1C 10 05 00 63 ec 0a 01 74 f5 cd 52 9d da 28 52 55 0d 94 f0 05 08 01 00 20 00 2a 00
+    |  |  |  |  \---------------------------------------------/ \--------/  \---/ |  |
+    |  |  |  |                ^-PlatformCapabilityUUID            |   wLength-^   |  |
+    |  |  |  ^-bReserved                                 dwVersion^   bVendorCode-^  |
+    |  |  ^-bDevCapability                                                bAltEnumCmd^
+    |  ^-bDescriptorType
+    ^-bLength
+
+The `dwVersion` encoded here is fwupd `1.8.5`, which is also the first version that supports BOS DS20 descriptors.
+A `dwVersion` of 0x0 will match any fwupd version, but should probably not be used unless required.
+
+The BOS descriptors are sorted by the requester and can appear in any order.
+The descriptor with the highest `dwVersion` that is not newer than the current running daemon version is used.
+This allows devices to set different quirks depending on the fwupd version, although in practice this should not be required.
```

Comment:
> What happens when a fix lands in 1.8.4 and backports to 1.7.8?

It's a good question, and one that's pretty common in everything that uses semvers and has multiple supported parallel development streams. The same can be said about RHEL 9.2, RHEL 8.7, etc. My gut instinct is we keep it simple and in the same way we don't backport plugins (in general) and we codify we don't backport features, only fixes.

Reply:
But so even if you only backport fixes you get caught up in this mess.  If there is a specific **fix** that you're depending upon for your device to upgrade then you can't be using semver on the backport anymore.

Reply:
I think in that case I'd rather support firmware level requirements, e.g. `fwupd > 1.8.2 OR (fwupd >= 1.7.9 AND fwupd < 1.8.0)`. The worst case for quirks is that the daemon will spit out a g_warning() I think. If it's worse than that (e.g. device hangs on custom control transfer) then we tell the vendor to only have the requirement on the new branch.

Reply:
OK so then if it's purely firmware metadata level requirements then I think this needs to be positioned differently.
Right now it's only probe if you're dwVersion 1.8.5 or later, but it should be able to probe no _differently_ 1.8.5, 1.9.0, or 2.0.0.

The key is whether it can match to a device and accept an update and that is tied entirely to the CAB then.

Reply:
I think the idea is that if you have a `foobarbaz` plugin added in 1.8.6 that then added a `FooBarBaz` quirk entry in 1.8.9, you'd have:

    dwVersion: 1.8.9
    data: Plugin=dfu\nFooBarBaz=whatever

    dwVersion: 1.8.5
    data: Plugin=foobarbaz

That way if you're using 1.8.9 or above then you'd use the first ds, and if you're using 1.8.6 you'd be using the second ds. If you're running 1.8.5 then you'd match neither.

Reply:
But aren't unknown quirks just ignored by older versions?

Reply:
> But aren't unknown quirks just ignored by older versions

Nah, there's a vocal g_warning() on the CLI -- which is probably the right thing to do to avoid typos.